File: backend/Services/ManufacturerService.cs
using MerxoSell.API.DTOs.Manufacturer;
using MerxoSell.API.Models;
using MerxoSell.API.Repositories.Interfaces;
using MerxoSell.API.Services.Interfaces;

namespace MerxoSell.API.Services;

public class ManufacturerService : IManufacturerService
{
    private readonly IManufacturerRepository _repo;

    public ManufacturerService(IManufacturerRepository repo) => _repo = repo;

    public async Task<IReadOnlyList<ManufacturerDto>> GetAllAsync()
    {
        var manufacturers = await _repo.GetAllAsync();
        return manufacturers.Select(Map).ToList();
    }

    public async Task<ManufacturerDto?> GetByIdAsync(int id)
    {
        var m = await _repo.GetByIdAsync(id);
        return m is null ? null : Map(m);
    }

    public async Task<ManufacturerDto> CreateAsync(CreateManufacturerDto dto)
    {
        var manufacturer = new Manufacturer
        {
            Name         = dto.Name,
            ContactEmail = dto.ContactEmail,
            Phone        = dto.Phone,
            Address      = dto.Address,
            Country      = dto.Country,
            Website      = dto.Website,
            IsActive     = true,
            CreatedAt    = DateTime.UtcNow
        };
        var created = await _repo.CreateAsync(manufacturer);
        return Map(created);
    }

    public async Task<ManufacturerDto?> UpdateAsync(int id, UpdateManufacturerDto dto)
    {
        var updated = await _repo.UpdateAsync(new Manufacturer
        {
            ManufacturerId = id,
            Name           = dto.Name,
            ContactEmail   = dto.ContactEmail,
            Phone          = dto.Phone,
            Address        = dto.Address,
            Country        = dto.Country,
            Website        = dto.Website,
            IsActive       = dto.IsActive
        });
        return updated is null ? null : Map(updated);
    }

    public Task<bool> DeleteAsync(int id) => _repo.DeleteAsync(id);

    private static ManufacturerDto Map(Manufacturer m) => new()
    {
        ManufacturerId = m.ManufacturerId,
        Name           = m.Name,
        ContactEmail   = m.ContactEmail,
        Phone          = m.Phone,
        Address        = m.Address,
        Country        = m.Country,
        Website        = m.Website,
        IsActive       = m.IsActive,
        CreatedAt      = m.CreatedAt,
        ProductCount   = m.Products.Count
    };
}

File: frontend/src/app/core/manufacturer/manufacturer.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface ManufacturerDto {
  manufacturerId: number;
  name: string;
  contactEmail?: string;
  phone?: string;
  address?: string;
  country?: string;
  website?: string;
  isActive: boolean;
  createdAt: string;
  productCount: number;
}

export interface CreateManufacturerDto {
  name: string;
  contactEmail?: string;
  phone?: string;
  address?: string;
  country?: string;
  website?: string;
}

export interface UpdateManufacturerDto extends CreateManufacturerDto {
  isActive: boolean;
}

@Injectable({ providedIn: 'root' })
export class ManufacturerService {
  private readonly base = `${environment.apiUrl}/manufacturers`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<ManufacturerDto[]> {
    return this.http.get<ManufacturerDto[]>(this.base);
  }

  getById(id: number): Observable<ManufacturerDto> {
    return this.http.get<ManufacturerDto>(`${this.base}/${id}`);
  }

  create(dto: CreateManufacturerDto): Observable<ManufacturerDto> {
    return this.http.post<ManufacturerDto>(this.base, dto);
  }

  update(id: number, dto: UpdateManufacturerDto): Observable<ManufacturerDto> {
    return this.http.put<ManufacturerDto>(`${this.base}/${id}`, dto);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}`);
  }
}
